modules/core/src/directx.cpp

```diff
@@ -1392,11 +1393,192 @@ void convertFromD3D11Texture2D(ID3D11Texture2D* pD3D11Texture2D, OutputArray dst
+        clGetDeviceIDsFromD3D11KHR_fn clGetDeviceIDsFromD3D11KHR = (clGetDeviceIDsFromD3D11KHR_fn)
+            clGetExtensionFunctionAddressForPlatform(platforms[i], "clGetDeviceIDsFromD3D11KHR");
+        if (!clGetDeviceIDsFromD3D11KHR) {
+            continue;
+        }
+        std::vector<std::pair<int, Device>> temp_devices;
+        // try with CL_PREFERRED_DEVICES_FOR_D3D11_KHR
+        temp_devices = findKHRDevice(clGetDeviceIDsFromD3D11KHR,
+                                     pD3D11Device,
+                                     CL_PREFERRED_DEVICES_FOR_D3D11_KHR,
+                                     platforms[i]);
+
+
+        // try with CL_ALL_DEVICES_FOR_D3D11_KHR
+        if (temp_devices.empty()) {
+            temp_devices = findKHRDevice(clGetDeviceIDsFromD3D11KHR,
+                                         pD3D11Device,
+                                         CL_ALL_DEVICES_FOR_D3D11_KHR,
+                                         platforms[i]);
+        }
+
+        if (!temp_devices.empty()) {
+            dst.insert(dst.end(), std::make_move_iterator(temp_devices.begin()),
+                                  std::make_move_iterator(temp_devices.end()));
+        }
+    }
+    return dst;
+#endif
+}
+
+std::tuple<Image2D, OpenCLExecutionContext> convertFromD3D11Texture2DtoCLMem(ID3D11Texture2D* pD3D11Texture2D, Device& device)
+{
+    CV_UNUSED(pD3D11Texture2D); CV_UNUSED(device);
+#if !defined(HAVE_DIRECTX)
+    NO_DIRECTX_SUPPORT_ERROR;
+#elif !defined(HAVE_OPENCL)
+    NO_OPENCL_SUPPORT_ERROR;
+#else
+    if (device.ptr() == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: convertFromD3D11Texture2DtoCLMem got empty Device object");
+    }
+
+    cl_device_id cl_device = static_cast<cl_device_id>(device.ptr());
+    size_t out_size = 0; // size in bytes
+    cl_int status = clGetDeviceInfo(cl_device, CL_DEVICE_PLATFORM, 0, nullptr, &out_size);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clGetDeviceInfo failed");
+    }
+    cl_platform_id platform = nullptr; // The platform associated with this device.
+    status = clGetDeviceInfo(cl_device, CL_DEVICE_PLATFORM, out_size, &platform, nullptr);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clGetDeviceInfo failed");
+    }
+    if (platform == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: clGetDeviceInfo returns empty platform");
+    }
+    auto&& platformName = PlatformInfo(&platform).name();
+
+    ID3D11Device *pD3D11Device = nullptr;
+    pD3D11Texture2D->GetDevice(&pD3D11Device);
+
+    if (pD3D11Device == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: GetDevice returns empty pD3D11Device in convertFromD3D11Texture2DtoCLMem");
+    }
+
+    if (!device.isExtensionSupported("cl_khr_d3d11_sharing")) {
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: Device doesn't support cl_khr_d3d11_sharing extension");
```

Comment:
don;t forget `pD3D11Device->Release();`
or move check extension upper before acquiring pD3D11Device: i suggest to move it in beginning of function

Reply:
Added